src/module_3/model_selection_converted.md

```diff
@@ -0,0 +1,524 @@
+
+Ridge regularization adds a penalty term to the coefficient estimates during model training, which helps prevent overfitting by reducing the magnitude of the coefficients. By starting with Ridge regularization, we ensure that no variables are overly penalized initially, allowing the model to learn from the data more effectively.
+
+To optimize the model, we aim to find the highest area under the precision-recall curve (AUC-PR) due to several key reasons:
+
+* Independent of Threshold: The AUC-PR metric remains independent of the threshold and focuses solely on the positive class, making it suitable for evaluating the model's performance in identifying potential purchasers regardless of the classification threshold. 
+
+* Sensitivity to Trade-offs: Our objective of increasing monthly sales by 2% and boosting selected items by 25% requires a balanced consideration of trade-offs between correctly identifying purchasers (recall) and minimizing false positives (precision). AUC-PR enables us to optimize this trade-off effectively, ensuring efficient resource allocation and maximizing the impact on sales performance.
+
+* Comprehensive Assessment: By encompassing the entire range of possible recall-precision trade-offs, AUC-PR offers a comprehensive evaluation of the model's performance. It captures the model's ability to correctly classify positive instances while also minimizing missed opportunities, aligning well with our overarching goal of improving sales performance and focusing on specific items for growth.
+
+
+```python
+score = metrics.make_scorer(metrics.average_precision_score, greater_is_better=True,needs_threshold=True)
+```
+
+
+```python
+pipe = Pipeline([
+    ('preprocessor', column_transformer),
+    ('scaler', StandardScaler()),
+    ('reg', LogisticRegression(solver='newton-cg', penalty='l2'))
+])
+
+
+reg_values = np.geomspace(0.00001, 1e+4, num=10)
+parameters = {'reg__C': 1.0/reg_values}
+
+gridsearch = GridSearchCV(pipe, parameters,  
+                          scoring=score, 
+                          return_train_score=True)
+
+gridsearch.fit(X_train, y_train)
+```
+
+
+```python
+results = pd.DataFrame(gridsearch.cv_results_)
+
+results = results[['param_reg__C', 'mean_test_score', 'mean_train_score']]
+
+results['param_reg__C'] = 1 / results['param_reg__C']
+
+results = results.rename(columns={'param_reg__C': 'Regularization', 'mean_test_score': 'Mean Test Score', 'mean_train_score': 'Mean Train Score'})
+results
+```
+
+The confusion matrix highlights the model's strong performance in correctly classifying non-purchases (true negatives), minimizing unnecessary notifications to users who are unlikely to engage with the product. However, the model struggles to accurately identify individuals who are likely to make a purchase (true positives), leading to missed opportunities for engagement and potential sales.
```

Comment:
Confusion matrix will use threshold = 0.5 which is meaningless. Cross validation can cause information leakage since it mixes past and future samples during the CV process.